flow/TLSConfig.actor.cpp

```diff
@@ -625,196 +633,365 @@ TLSPolicy::Rule::Rule(std::string input) {
+		return (criteria_utf8_len == entry_utf8_len &&
+		        memcmp(criteria_utf8.get(), entry_utf8.get(), criteria_utf8_len) == 0);
+	case MatchType::PREFIX:
+		return (criteria_utf8_len <= entry_utf8_len &&
+		        memcmp(criteria_utf8.get(), entry_utf8.get(), criteria_utf8_len) == 0);
+	case MatchType::SUFFIX:
+		return (criteria_utf8_len <= entry_utf8_len && memcmp(criteria_utf8.get(),
+		                                                      entry_utf8.get() + (entry_utf8_len - criteria_utf8_len),
+		                                                      criteria_utf8_len) == 0);
+	default:
+		UNREACHABLE();
+	}
 }
 
-bool match_extension_criteria(X509* cert, NID nid, const std::string& value, MatchType mt) {
-	if (nid != NID_subject_alt_name && nid != NID_issuer_alt_name) {
-		// I have no idea how other extensions work.
-		return false;
+class PeerVerifier {
+	const std::vector<TLSPolicy::Rule>& m_rules;
+	X509_STORE_CTX* m_storeCtx;
+	const bool m_isClient;
+	bool m_verified;
+
+	std::string_view m_successReason;
+
+	std::vector<std::string_view> m_verifyState;
+	std::string m_currentName;
+	std::string m_currentExtension;
+	std::vector<std::string> m_failureReasons;
+
+	bool matchNameCriteria(X509_NAME* name, const NID nid, std::string_view criteria, const MatchType matchType);
+
+	bool matchExtensionCriteria(const X509* cert, const NID nid, std::string_view criteria, const MatchType matchType);
+
+	bool matchCriteria(const X509* cert, X509_NAME* name, const NID nid, const Criteria& criteria);
+
+	// Verify a set of criteria
+	bool verifyCriterias(const X509* cert, X509_NAME* name, const TLSPolicy::Rule::CriteriaMap& criterias);
+
+	// Verify a single rule
+	bool verifyRule(const TLSPolicy::Rule&);
+
+	// Verify the TLSPolicy peer
+	bool verify();
+
+public:
+	PeerVerifier(const std::vector<TLSPolicy::Rule>& rules, X509_STORE_CTX* store_ctx, const bool is_client)
+	  : m_rules(rules), m_storeCtx(store_ctx), m_isClient(is_client), m_successReason(), m_verifyState() {
+		ASSERT(m_storeCtx != nullptr);
+
+		// Prealloc 32 * sizeof(const char*) to avoid reallocation, this should be sufficient
+		m_verifyState.reserve(32);
+
+		m_verified = verify();
+	}
+
+	bool isOk() const noexcept { return m_verified; }
+	bool isErr() const noexcept { return !isOk(); }
+	const std::vector<std::string>& getFailureReasons() const noexcept { return m_failureReasons; }
+	const std::string_view getSuccessReason() const noexcept { return m_successReason; }
+};
+
+std::string getX509Name(const X509_NAME* name) {
+	std::unique_ptr<BIO, decltype(&BIO_free)> out(BIO_new(BIO_s_mem()), BIO_free);
+	if (out == nullptr) {
+		return std::string("Unable to allocate OpenSSL BIO");
 	}
-	auto pos = value.find(':');
-	if (pos == value.npos) {
+	X509_NAME_print_ex(out.get(), name, /* indent= */ 0, /* flags */ XN_FLAG_ONELINE);
+	unsigned char* rawName = nullptr;
+	int length = BIO_get_mem_data(out.get(), &rawName);
+	std::string result = (char*)rawName;
```

Comment:
```suggestion
	std::string result((char*)rawName, length);
```
And then we can remove the next assertion line.

Reply:
I think this assertion is intentional. One reason is that, `BIO_get_mem_data` does not enforce a `\0` terminating name. The other reason is taht `BIO_get_mem_data` will return the total amount of data available. I have no detailed understanding about `BIO` data structure, so I would guess it would be safe to set an `ASSERT_EQ` here and see what happens.